### PR TITLE
bug/5821-bounding-boxes-are-offset

### PR DESCRIPTION
**Pull request by briemla**
OP#5821
OP#6078

### CODE REVIEW

File: OTAnalytics/plugin_video_processing/video_reader.py [hunk 1]
```diff
@@ -1,52 +1,78 @@
...
+                raise InvalidVideoError(f"{str(video_path)} is not a video")
+            video = container.streams.video[0]
+            max_frames = video.frames
+            frame_to_read = min(frame_to_read, max_frames - OFFSET)
+            framerate = self.__get_fps(container, video_path)
+            time_base = (
+                video.time_base if video.time_base else Fraction(av.time_base, 1)
+            )
+            time_in_video = int(frame_to_read / framerate)
+            container.seek(time_in_video * av.time_base, backward=True)
```
Comment (randy-seng) [inline OTAnalytics/plugin_video_processing/video_reader.py:65]: Is there a reason why we are not using time_base that we declared above?

File: tests/OTAnalytics/plugin_video_processing/test_video_reader.py [hunk 2]
```diff
@@ -1,24 +1,58 @@
...
+        self, cyclist_video: Path, video_reader: VideoReader
+    ) -> None:
         delta = timedelta(seconds=1)
 
-        frame_number = self.video_reader.get_frame_number_for(cyclist_video, delta)
+        frame_number = video_reader.get_frame_number_for(cyclist_video, delta)
 
         assert frame_number == 20
+
+    @pytest.mark.parametrize("frame_num", (1, 5))
```
Comment (randy-seng) [inline tests/OTAnalytics/plugin_video_processing/test_video_reader.py:49]: Assuming the video has 60 frames, can we test the edge cases 0, 60, 61?

**Review by randy-seng — CHANGES_REQUESTED** [review]
I have a couple of questions~

File: OTAnalytics/plugin_video_processing/video_reader.py [hunk 3]
```diff
@@ -1,52 +1,78 @@
...
+        frame = self._read_frame(frame_number, video_path)
+        return av_to_image(frame)
+
+    def _read_frame(self, frame_to_read: int, video_path: Path) -> VideoFrame:
+        with self.__get_clip(video_path) as container:
+            if len(container.streams.video) <= 0:
+                raise InvalidVideoError(f"{str(video_path)} is not a video")
+            video = container.streams.video[0]
+            max_frames = video.frames
+            frame_to_read = min(frame_to_read, max_frames - OFFSET)
```
[the remark this answers is not in the document]
Comment (briemla) [reply OTAnalytics/plugin_video_processing/video_reader.py:59]: A video with `n` frames uses the frame numbers `0` to `n-1`. `frame_to_read` is used twice in here. First, it is used to calculate the seek time within the video (time to jump directly to aka `time_in_video`). Second, it is used to iterate to the correct frame after `time_in_video` or in frames `sec_frame`. To calculate `sec_frame` one frame will be read (line 66). Therefore, to get to frame `frame_to_read` given as parameter, the maximum is `n - 2`.

File: tests/OTAnalytics/plugin_video_processing/test_video_reader.py (hunk 2 above)
Comment (briemla) [reply tests/OTAnalytics/plugin_video_processing/test_video_reader.py]: I added the edge cases: -1, 0, 58, 59, 60 and 61

File: OTAnalytics/plugin_video_processing/video_reader.py (hunk 1 above)
Comment (briemla) [reply OTAnalytics/plugin_video_processing/video_reader.py]: Yes, I refactored the code and added some documentation.

**Review by randy-seng — APPROVED** [review]
Looks good!
